File: gpt-4-turbo/stage2/kernels_construct_reduction_clause.cpp
#include <iostream>
#include <cstdlib>
#include <cmath>

// Define SEED for reproducibility
#define SEED 123
// Define the number of elements in the array and number of test calls
#define N 1024
#define NUM_TEST_CALLS 10

int test1(){
    int err = 0;
    srand(SEED);

    // Create and initialize an array with random values
    double *a = new double[N];
    for (int i = 0; i < N; ++i) {
        a[i] = rand() % 100;  // Random values between 0 and 99
    }

    double reduction_result = 0;
    double expected_result = 0;

    // Calculate expected_result using CPU
    for (int i = 0; i < N; ++i){
        expected_result += a[i];
    }

    // Using OpenACC to perform reduction on the array
#pragma acc kernels copy(reduction_result) create(a[:N]) reduction(+:reduction_result)
    {
        for (int i = 0; i < N; ++i) {
            reduction_result += a[i];
        }
    }

    // Verifying the result of the OpenACC reduction with the expected CPU result
    if (fabs(reduction_result - expected_result) > 1e-5) { // Using a tolerance to account for floating-point arithmetic inaccuracies
        err++;
    }

    // Cleanup
    delete[] a;

    return err;
}

int main(){
    int failcode = 0;
    int failed;

    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test1();
    }
    if (failed != 0){
        failcode = failcode + (1 << 0);
    }

    if (failcode == 0) {
        std::cout << "PASS: OpenACC kernels construct reduction clause works as expected." << std::endl;
    } else {
        std::cout << "FAIL: OpenACC kernels construct reduction clause does not work as expected. Failures: " << failed << std::endl;
    }

    return failcode;
}

File: gpt-4-turbo/stage2/kernels_construct_reduction_clause.f90
!T1:kernels construct reduction clause,V:2.7-2.3
      LOGICAL FUNCTION test1()
        USE OPENACC
        IMPLICIT NONE
        INCLUDE "acc_testsuite.Fh"
        
        INTEGER, PARAMETER :: N = 1024
        REAL :: a(N)
        REAL :: total, expected
        INTEGER :: i,errors

        errors = 0
        total = 0.0
        expected = 0.0

        ! Initialize a
        DO i = 1, N
           a(i) = 1.0
           expected = expected + a(i)
        END DO

        ! Apply OpenACC kernels reduction
        !$acc kernels reduction(+:total)
        DO i = 1, N
           total = total + a(i)
        END DO
        !$acc end kernels

        ! Check if the reduction result is correct
        IF (ABS(total - expected) > 1.0E-5) THEN
           errors = errors + 1
        END IF

        IF (errors .eq. 0) THEN
          test1 = .FALSE.
        ELSE
          test1 = .TRUE.
        END IF
      END

      PROGRAM main
        IMPLICIT NONE
        INTEGER :: failcode, testrun
        LOGICAL :: failed
        INCLUDE "acc_testsuite.Fh"
        EXTERNAL test1
        failed = .FALSE.
        failcode = 0
        DO testrun = 1, NUM_TEST_CALLS
          failed = failed .or. test1()
        END DO
        IF (failed) THEN
          failcode = failcode + 2 ** 0
          failed = .FALSE.
        END IF
        CALL EXIT (failcode)
      END PROGRAM